Multiple test failures on Windows

OS: Win7 x64
Go: 1.5.1

Pulling this out of https://github.com/constabulary/gb/pull/377 per request.

```
C:\Users\alan\gopath\src\github.com\constabulary\gb>go test -short github.com/co
nstabulary/gb/...
a
a
b
a
d.v1
c
d.v1
cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
notestfiles
cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
--- FAIL: TestBuild (0.62s)
        build_test.go:74: ctx.Build(cgomain): want <nil>, got # C:\Go\pkg\tool\w
indows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb281314287\cgomain\_c
go -importpath cgomain -- -I C:\Users\alan\AppData\Local\Temp\gb281314287\cgomai
n\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cgoma
in main.go: exit status 2
        build_test.go:74: ctx.Build(cgotest): want <nil>, got # C:\Go\pkg\tool\w
indows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb997494402\cgotest\_c
go -importpath cgotest -- -I C:\Users\alan\AppData\Local\Temp\gb997494402\cgotes
t\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cgote
st basic.go: exit status 2
        build_test.go:74: ctx.Build(cgoonlynotest): want <nil>, got # C:\Go\pkg\
tool\windows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb557978884\cgoo
nly\_cgo -importpath cgoonlynotest -- -I C:\Users\alan\AppData\Local\Temp\gb5579
78884\cgoonly\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testda
ta\src\cgoonlynotest cgoonly.go: exit status 2
        build_test.go:74: ctx.Build(cppmain): want <nil>, got # C:\Go\pkg\tool\w
indows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb119974520\cppmain\_c
go -importpath cppmain -- -I C:\Users\alan\AppData\Local\Temp\gb119974520\cppmai
n\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cppma
in main.go: exit status 2
a
a
b
a
d.v1
c
d.v1
a
b
--- FAIL: TestPackageBinfile (0.00s)
        package_test.go:68: (b).Binfile(): want C:\Users\alan\gopath\src\github.
com\constabulary\gb\testdata\bin\b, got C:\Users\alan\gopath\src\github.com\cons
tabulary\gb\testdata\bin\b.exe
FAIL
FAIL    github.com/constabulary/gb      1.409s
*** Test killed: ran too long (10m0s).
FAIL    github.com/constabulary/gb/cmd  600.170s
ok      github.com/constabulary/gb/cmd/gb       0.100s
?       github.com/constabulary/gb/cmd/gb-vendor        [no test files]
?       github.com/constabulary/gb/log  [no test files]
ok      github.com/constabulary/gb/vendor       0.245s
```

What gcc do you have on your test machine ?

On Thu, Sep 17, 2015 at 4:58 PM, Alan Shreve <notifications@github.com>
wrote:

> OS: Win7 x64
> Go: 1.5.1
>
> Pulling this out of #377 <https://github.com/constabulary/gb/pull/377>
> per request.
>
> C:\Users\alan\gopath\src\github.com\constabulary\gb>go test -short github.com/co
> nstabulary/gb/. <http://github.com/constabulary/gb/.>..
> a
> a
> b
> a
> d.v1
> c
> d.v1
> cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
> cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
> notestfiles
> cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
> cc1.exe: sorry, unimplemented: 64-bit mode not compiled in
> --- FAIL: TestBuild (0.62s)
>         build_test.go:74: ctx.Build(cgomain): want <nil>, got # C:\Go\pkg\tool\w
> indows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb281314287\cgomain\_c
> go -importpath cgomain -- -I C:\Users\alan\AppData\Local\Temp\gb281314287\cgomai
> n\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cgoma
> in main.go: exit status 2
>         build_test.go:74: ctx.Build(cgotest): want <nil>, got # C:\Go\pkg\tool\w
> indows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb997494402\cgotest\_c
> go -importpath cgotest -- -I C:\Users\alan\AppData\Local\Temp\gb997494402\cgotes
> t\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cgote
> st basic.go: exit status 2
>         build_test.go:74: ctx.Build(cgoonlynotest): want <nil>, got # C:\Go\pkg\
> tool\windows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb557978884\cgoo
> nly\_cgo -importpath cgoonlynotest -- -I C:\Users\alan\AppData\Local\Temp\gb5579
> 78884\cgoonly\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testda
> ta\src\cgoonlynotest cgoonly.go: exit status 2
>         build_test.go:74: ctx.Build(cppmain): want <nil>, got # C:\Go\pkg\tool\w
> indows_amd64\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb119974520\cppmain\_c
> go -importpath cppmain -- -I C:\Users\alan\AppData\Local\Temp\gb119974520\cppmai
> n\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cppma
> in main.go: exit status 2
> a
> a
> b
> a
> d.v1
> c
> d.v1
> a
> b
> --- FAIL: TestPackageBinfile (0.00s)
>         package_test.go:68: (b).Binfile(): want C:\Users\alan\gopath\src\github.
> com\constabulary\gb\testdata\bin\b, got C:\Users\alan\gopath\src\github.com\cons
> tabulary\gb\testdata\bin\b.exe
> FAIL
> FAIL    github.com/constabulary/gb      1.409s
> *** Test killed: ran too long (10m0s).
> FAIL    github.com/constabulary/gb/cmd  600.170s
> ok      github.com/constabulary/gb/cmd/gb       0.100s
> ?       github.com/constabulary/gb/cmd/gb-vendor        [no test files]
> ?       github.com/constabulary/gb/log  [no test files]
> ok      github.com/constabulary/gb/vendor       0.245s
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/constabulary/gb/issues/384>.
>

not entirely sure what you're asking here? there's not GOARCH environment variable set, so it's using the OS default (amd64). running the tests with GOARCH=386 produces additional tests errors:

```
C:\Users\alan\gopath\src\github.com\constabulary\gb>set GOARCH=386

C:\Users\alan\gopath\src\github.com\constabulary\gb>go test -short github.com/co
nstabulary/gb/...
a
a
a
d.v1
d.v1
notestfiles
--- FAIL: TestBuild (0.03s)
        build_test.go:74: ctx.Build(a): want <nil>, got exec: "C:\\Go\\pkg\\tool
\\windows_386\\compile.exe": file does not exist
        build_test.go:74: ctx.Build(b): want <nil>, got exec: "C:\\Go\\pkg\\tool
\\windows_386\\compile.exe": file does not exist
        build_test.go:74: ctx.Build(c): want <nil>, got exec: "C:\\Go\\pkg\\tool
\\windows_386\\compile.exe": file does not exist
        build_test.go:74: ctx.Build(d.v1): want <nil>, got exec: "C:\\Go\\pkg\\t
ool\\windows_386\\compile.exe": file does not exist
        build_test.go:74: ctx.Build(cgomain): want <nil>, got # C:\Go\pkg\tool\w
indows_386\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb642678559\cgomain\_cgo
 -importpath cgomain -- -I C:\Users\alan\AppData\Local\Temp\gb642678559\cgomain\
_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cgomain
 main.go: exec: "C:\\Go\\pkg\\tool\\windows_386\\cgo": file does not exist
        build_test.go:74: ctx.Build(cgotest): want <nil>, got # C:\Go\pkg\tool\w
indows_386\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb116975602\cgotest\_cgo
 -importpath cgotest -- -I C:\Users\alan\AppData\Local\Temp\gb116975602\cgotest\
_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cgotest
 basic.go: exec: "C:\\Go\\pkg\\tool\\windows_386\\cgo": file does not exist
        build_test.go:74: ctx.Build(notestfiles): want <nil>, got exec: "C:\\Go\
\pkg\\tool\\windows_386\\compile.exe": file does not exist
        build_test.go:74: ctx.Build(cgoonlynotest): want <nil>, got # C:\Go\pkg\
tool\windows_386\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb296190964\cgoonl
y\_cgo -importpath cgoonlynotest -- -I C:\Users\alan\AppData\Local\Temp\gb296190
964\cgoonly\_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata
\src\cgoonlynotest cgoonly.go: exec: "C:\\Go\\pkg\\tool\\windows_386\\cgo": file
 does not exist
        build_test.go:74: ctx.Build(cppmain): want <nil>, got # C:\Go\pkg\tool\w
indows_386\cgo -objdir C:\Users\alan\AppData\Local\Temp\gb922554984\cppmain\_cgo
 -importpath cppmain -- -I C:\Users\alan\AppData\Local\Temp\gb922554984\cppmain\
_cgo -I C:\Users\alan\gopath\src\github.com\constabulary\gb\testdata\src\cppmain
 main.go: exec: "C:\\Go\\pkg\\tool\\windows_386\\cgo": file does not exist
a
a
a
d.v1
--- FAIL: TestExecuteBuildAction (0.01s)
        executor_test.go:53: Execute(build: a): want: <nil>, got exec: "C:\\Go\\
pkg\\tool\\windows_386\\compile.exe": file does not exist
        executor_test.go:53: Execute(build: b): want: <nil>, got exec: "C:\\Go\\
pkg\\tool\\windows_386\\compile.exe": file does not exist
        executor_test.go:53: Execute(build: c): want: <nil>, got exec: "C:\\Go\\
pkg\\tool\\windows_386\\compile.exe": file does not exist
        executor_test.go:53: Execute(build: d.v1): want: <nil>, got exec: "C:\\G
o\\pkg\\tool\\windows_386\\compile.exe": file does not exist
a
--- FAIL: TestStale (0.00s)
        install_test.go:76: exec: "C:\\Go\\pkg\\tool\\windows_386\\compile.exe":
 file does not exist
--- FAIL: TestPackageBinfile (0.00s)
        package_test.go:68: (b).Binfile(): want C:\Users\alan\gopath\src\github.
com\constabulary\gb\testdata\bin\b, got C:\Users\alan\gopath\src\github.com\cons
tabulary\gb\testdata\bin\b.exe
FAIL
FAIL    github.com/constabulary/gb      0.180s
```

Awesome. Yeah, just to confirm, it's this timeout bug:

```
C:\Users\alan\gopath>go test -short -timeout=15s github.com\constabulary\gb\...
ok      github.com/constabulary/gb      5.739s
panic: test timed out after 15s

goroutine 12 [running]:
testing.startAlarm.func1()
        c:/go/src/testing/testing.go:703 +0x139
created by time.goFunc
        c:/go/src/time/sleep.go:129 +0x41

goroutine 1 [chan receive]:
testing.RunTests(0x7392b8, 0x812420, 0xb, 0xb, 0xc08202f401)
        c:/go/src/testing/testing.go:562 +0x8b4
testing.(*M).Run(0xc082081ef8, 0x5)
        c:/go/src/testing/testing.go:494 +0x77
main.main()
        github.com/constabulary/gb/cmd/_test/_testmain.go:74 +0x11d

goroutine 9 [runnable]:
runtime.Gosched()
        c:/go/src/runtime/proc.go:166 +0x14
path/filepath.Clean(0xc082465ff2, 0x5, 0x0, 0x0)
        c:/go/src/path/filepath/path.go:145 +0x84d
path/filepath.joinNonEmpty(0xc082085c58, 0x2, 0x2, 0x0, 0x0)
        c:/go/src/path/filepath/path_windows.go:126 +0x9d
path/filepath.join(0xc082085c58, 0x2, 0x2, 0x0, 0x0)
        c:/go/src/path/filepath/path_windows.go:115 +0xd9
path/filepath.Join(0xc082085c58, 0x2, 0x2, 0x0, 0x0)
        c:/go/src/path/filepath/path.go:206 +0x48
github.com/constabulary/gb/cmd.FindProjectroot(0xc082465fd0, 0x3, 0x0, 0x0, 0x0,
 0x0)
        C:/Users/alan/gopath/src/github.com/constabulary/gb/cmd/path.go:20 +0x19
1
github.com/constabulary/gb/cmd.TestFindProjectroot(0xc082094120)
        C:/Users/alan/gopath/src/github.com/constabulary/gb/cmd/path_test.go:58
+0x7cd
testing.tRunner(0xc082094120, 0x812450)
        c:/go/src/testing/testing.go:456 +0x9f
created by testing.RunTests
        c:/go/src/testing/testing.go:561 +0x874
FAIL    github.com/constabulary/gb/cmd  15.295s
ok      github.com/constabulary/gb/cmd/gb       0.208s
?       github.com/constabulary/gb/cmd/gb-vendor        [no test files]
?       github.com/constabulary/gb/log  [no test files]
ok      github.com/constabulary/gb/vendor       0.234s
```

--- a/cmd/path.go
+++ b/cmd/path.go
@@ -16,7 +16,7 @@ func FindProjectroot(path string) (string, error) {
 		return "", fmt.Errorf("project root is blank")
 	}
 	start := path
-	for path != "/" {
+	for path != filepath.Dir(path) {
 		root := filepath.Join(path, "src")
 		if _, err := os.Stat(root); err != nil {
 			if os.IsNotExist(err) {

--- a/cmd/test_test.go
+++ b/cmd/test_test.go
@@ -1,6 +1,7 @@
 package cmd
 
 import (
+	"path/filepath"
 	"reflect"
 	"sort"
 	"strings"
@@ -171,7 +172,7 @@ func TestTestPackages(t *testing.T) {
 		sort.Strings(actual)
 		var expected []string
 		for _, s := range tt.actions {
-			expected = append(expected, strings.Replace(s, "$WORKDIR", ctx.Workdir(), -1))
+			expected = append(expected, strings.Replace(filepath.FromSlash(s), "$WORKDIR", ctx.Workdir(), -1))
 		}
 		if !reflect.DeepEqual(expected, actual) {
 			t.Errorf("TestBuildPackages(%v): want %v, got %v", pkgs, expected, actual)

--- a/cmd/test_test.go
+++ b/cmd/test_test.go
@@ -143,10 +143,10 @@ func TestTestPackages(t *testing.T) {
 		err     error
 	}{{
 		pkgs:    []string{"a", "b", "c"},
-		actions: []string{"run: [$WORKDIR/a/testmain/_test/a.test]", "run: [$WORKDIR/b/testmain/_test/b.test]", "run: [$WORKDIR/c/testmain/_test/c.test]"},
+		actions: []string{"run: [$WORKDIR/a/testmain/_test/a$EXE.test]", "run: [$WORKDIR/b/testmain/_test/b$EXE.test]", "run: [$WORKDIR/c/testmain/_test/c$EXE.test]"},
 	}, {
 		pkgs:    []string{"cgotest", "cgomain", "notestfiles", "cgoonlynotest", "testonly", "extestonly"},
-		actions: []string{"run: [$WORKDIR/cgomain/testmain/_test/cgomain.test]", "run: [$WORKDIR/cgoonlynotest/testmain/_test/cgoonly.test]", "run: [$WORKDIR/cgotest/testmain/_test/cgotest.test]", "run: [$WORKDIR/extestonly/testmain/_test/extestonly.test]", "run: [$WORKDIR/notestfiles/testmain/_test/notest.test]", "run: [$WORKDIR/testonly/testmain/_test/testonly.test]"},
+		actions: []string{"run: [$WORKDIR/cgomain/testmain/_test/cgomain$EXE.test]", "run: [$WORKDIR/cgoonlynotest/testmain/_test/cgoonly$EXE.test]", "run: [$WORKDIR/cgotest/testmain/_test/cgotest$EXE.test]", "run: [$WORKDIR/extestonly/testmain/_test/extestonly$EXE.test]", "run: [$WORKDIR/notestfiles/testmain/_test/notest$EXE.test]", "run: [$WORKDIR/testonly/testmain/_test/testonly$EXE.test]"},
 	}}
 
 	for _, tt := range tests {
@@ -171,8 +171,15 @@ func TestTestPackages(t *testing.T) {
 		}
 		sort.Strings(actual)
 		var expected []string
+		exe := ""
+		if ctx.Context.GOOS == "windows" {
+			exe = ".exe"
+		}
 		for _, s := range tt.actions {
-			expected = append(expected, strings.Replace(filepath.FromSlash(s), "$WORKDIR", ctx.Workdir(), -1))
+			s = filepath.FromSlash(s)
+			s = strings.Replace(s, "$WORKDIR", ctx.Workdir(), -1)
+			s = strings.Replace(s, "$EXE", exe, -1)
+			expected = append(expected, s)
 		}
 		if !reflect.DeepEqual(expected, actual) {
 			t.Errorf("TestBuildPackages(%v): want %v, got %v", pkgs, expected, actual)

